app/controllers/declaration/AdditionalProcedureCodesController.scala

```diff
@@ -183,14 +191,22 @@ class AdditionalProcedureCodesController @Inject()(
     updateExportsDeclarationSyncDirect(updatedModel(_))
   }
 
-  private def nextPage(action: FormAction, mode: Mode, itemId: String, maybeProcedureCode: Option[String])(
+  private def nextPage(action: FormAction, mode: Mode, itemId: String, maybeProcedureCode: Option[String], maybeModel: Option[ExportsDeclaration])(
     implicit request: JourneyRequest[AnyContent]
-  ) =
+  ): Future[Result] =
     (action, maybeProcedureCode) match {
-      case (Add, _) => navigator.continueTo(mode, routes.AdditionalProcedureCodesController.displayPage(_, itemId))
+      case (Add, _) =>
+        Future.successful(navigator.continueTo(mode, routes.AdditionalProcedureCodesController.displayPage(_, itemId)))
+
       case (_, Some(code)) if ProcedureCodesData.osrProcedureCodes.contains(code) =>
-        navigator.continueTo(mode, routes.FiscalInformationController.displayPage(_, itemId))
-      case _ => navigator.continueTo(mode, routes.CommodityDetailsController.displayPage(_, itemId))
+        Future.successful(navigator.continueTo(mode, routes.FiscalInformationController.displayPage(_, itemId)))
+
+      case _ =>
+        val continueToCommodityDetails = navigator.continueTo(mode, routes.CommodityDetailsController.displayPage(_, itemId))
+
+        val model = maybeModel.fold(request.cacheModel)(identity)
+        if (SupervisingCustomsOfficeHelper.isConditionForAllProcedureCodesNotVerified(model)) Future.successful(continueToCommodityDetails)
+        else SupervisingCustomsOfficeHelper.resetCache(this).map(_ => continueToCommodityDetails)
```

Comment:
Would this logic not be better placed inside the `updateCache` method (it has no effect on which page we continue to)?

Reply:
`updateCache` is not the right place because it is called for every `ExportItem`.
The condition for resetting `supervisingCustomsOffice` is instead that **ALL** `ExportItem`s have 1040-000 as PC and APC.
I could end-up resetting the value just after the 1st `ExportItem` with 1040-000. A 2nd one could not have 1040-000 as PC-APC and in that case the condition will not be verified.

Reply:
Ah yes, you mean it is called for every new APC that gets added/removed (before continuing to the next page) :+1: